test/retain-history/mzcompose.py

```diff
@@ -0,0 +1,208 @@
+            1 100
+            2 200
+
+            > SELECT * FROM retain_history_mv_on_mv1 AS OF '{mz_time1}'::TIMESTAMP;
+            1 100
+            2 200
+
+            > SELECT * FROM retain_history_mv1 AS OF '{mz_time2}'::TIMESTAMP;
+            1 101
+            2 201
+            3 300
+            4 400
+
+            > SELECT * FROM retain_history_mv1 AS OF AT LEAST '{mz_time2}'::TIMESTAMP;
+            1 101
+            2 201
+            3 300
+            4 400
+
+            > SELECT sum(value), max(value) FROM retain_history_mv1 AS OF '{mz_time2}'::TIMESTAMP;
+            1002 400
+
+            > SELECT count(*) FROM retain_history_mv1 AS OF '{mz_time3}'::TIMESTAMP;
+            2
+
+            ? EXPLAIN SELECT * FROM retain_history_mv1 AS OF '{mz_time2}'::TIMESTAMP;
+            Explained Query:
+              ReadStorage materialize.public.retain_history_mv1
+
+            > SELECT mv1a.key, mv1b.key
+              FROM retain_history_mv1 mv1a
+              LEFT OUTER JOIN retain_history_mv1 mv1b
+              ON mv1a.key = mv1b.key
+              AS OF '{mz_time2}'::TIMESTAMP;
+            1 1
+            2 2
+            3 3
+            4 4
+
+            ! SELECT t.key, mv.key
+              FROM retain_history_table t
+              LEFT OUTER JOIN retain_history_mv1 mv
+              ON t.key = mv.key
+              AS OF '{mz_time2}'::TIMESTAMP;
+            contains: is not valid for all inputs
+
+            > UPDATE retain_history_table SET key = 9 WHERE key = 1;
+            """,
+    )
+
+    mz_time4 = fetch_now_from_mz(c)
+
+    testdrive(
+        c,
+        f"""
+            > SELECT count(*) FROM retain_history_mv1 WHERE key = 1 AS OF '{mz_time3}'::TIMESTAMP;
+            1
+
+            > SELECT count(*) FROM retain_history_mv1 WHERE key = 1 AS OF '{mz_time4}'::TIMESTAMP;
+            0
+
+            > SELECT 1 WHERE 1 = (SELECT count(*) FROM retain_history_mv1 WHERE key = 1) AS OF '{mz_time3}'::TIMESTAMP;
+            1
+            """,
+    )
+
+    test_time5 = datetime.now()
+
+    if (test_time5 - test_time1).total_seconds() <= mv_on_mv1_retention_in_sec:
+        time.sleep(1)
+
+    assert (
+        test_time5 - test_time1
+    ).total_seconds() < mv_on_mv_on_mv1_retention_in_sec, "test precondition not satisfied, consider increasing 'mv_on_mv_on_mv1_retention_in_sec'"
+
+    testdrive(
+        c,
+        f"""
+            # retain period exceeded
+            ! SELECT * FROM retain_history_mv_on_mv1 AS OF '{mz_time1}'::TIMESTAMP;
+            contains: is not valid for all inputs
+
+            # retain period on wrapping mv still valid
+            > SELECT * FROM retain_history_mv_on_mv_on_mv1 AS OF '{mz_time1}'::TIMESTAMP;
+            1 100
+            2 200
+            """,
+    )
+
+
+def testdrive(c: Composition, sql: str) -> None:
+    c.testdrive(
+        dedent(sql),
+        default_timeout="5s",
+        no_reset=True,
```

Comment:
Why not create `Testdrive()` with `no_reset=True`?

Reply:
Valid point! Adjusted and inlined.